File: python/api/project.py
from python.helpers.api import ApiHandler, Request, Response
from python.helpers.project import ProjectHelper, ProjectData
from python.helpers.print_style import PrintStyle
from flask import jsonify
from typing import Dict, Any


class Project(ApiHandler):
    """
    API handler for project management operations.
    
    Supports creating, listing, loading, activating, and deactivating projects
    through HTTP POST requests with JSON or form data.
    """

    async def process(self, input: dict, request: Request) -> dict | Response:
        """
        Process project management API requests
        
        Args:
            input: Input data from request (JSON or form data)
            request: Flask request object
            
        Returns:
            Response dict or Flask Response object
        """
        try:
            # Handle both JSON and multipart/form-data
            if request.content_type.startswith("multipart/form-data"):
                action = request.form.get("action", "").lower().strip()
                name = request.form.get("name", "").strip()
                description = request.form.get("description", "").strip()
                instructions = request.form.get("instructions", "").strip()
                ctxid = request.form.get("context", "")
            else:
                # Handle JSON request
                action = input.get("action", "").lower().strip()
                name = input.get("name", "").strip()
                description = input.get("description", "").strip()
                instructions = input.get("instructions", "").strip()
                ctxid = input.get("context", "")

            # Get agent context if needed for activate/deactivate operations
            context = None
            if action in ["activate", "deactivate", "get_active"]:
                context = self.get_context(ctxid)

            # Route to appropriate handler
            if action == "create":
                return await self._handle_create(name, description, instructions)
            elif action == "list":
                return await self._handle_list()
            elif action == "load":
                return await self._handle_load(name)
            elif action == "activate":
                return await self._handle_activate(context, name)
            elif action == "deactivate":
                return await self._handle_deactivate(context)
            elif action == "get_active":
                return await self._handle_get_active(context)
            elif action == "update":
                return await self._handle_update(name, description, instructions)
            else:
                return {
                    "success": False,
                    "error": f"Unknown action '{action}'. Available actions: create, list, load, activate, deactivate, get_active, update"
                }

        except Exception as e:
            error_msg = f"Project API error: {str(e)}"
            PrintStyle(font_color="red", padding=True).print(error_msg)
            return {
                "success": False,
                "error": error_msg
            }

    async def _handle_create(self, name: str, description: str, instructions: str) -> Dict[str, Any]:
        """Handle project creation"""
        if not name:
            return {
                "success": False,
                "error": "Project name is required"
            }

        result = ProjectHelper.create_project(
            name=name,
            description=description,
            instructions=instructions
        )

        return result

    async def _handle_list(self) -> Dict[str, Any]:
        """Handle listing all projects"""
        try:
            projects = ProjectHelper.list_projects()
            
            project_list = []
            for project in projects:
                project_list.append({
                    "name": project.name,
                    "description": project.description,
                    "instructions": project.instructions,
                    "directory": ProjectHelper.get_project_directory(project.name)
                })
            
            return {
                "success": True,
                "projects": project_list
            }

        except Exception as e:
            return {
                "success": False,
                "error": f"Failed to list projects: {str(e)}"
            }

    async def _handle_load(self, name: str) -> Dict[str, Any]:
        """Handle loading a specific project"""
        if not name:
            return {
                "success": False,
                "error": "Project name is required"
            }

        project = ProjectHelper.load_project(name)
        
        if not project:
            return {
                "success": False,
                "error": f"Project '{name}' not found"
            }

        return {
            "success": True,
            "project": {
                "name": project.name,
                "description": project.description,
                "instructions": project.instructions,
                "directory": ProjectHelper.get_project_directory(project.name),
                "file_structure": ProjectHelper.get_project_file_structure(project.name)
            }
        }

    async def _handle_activate(self, context, name: str) -> Dict[str, Any]:
        """Handle project activation"""
        if not context:
            return {
                "success": False,
                "error": "Agent context is required for project activation"
            }

        if not name:
            return {
                "success": False,
                "error": "Project name is required"
            }

        success = ProjectHelper.set_active_project(context.agent0, name)
        
        if success:
            project = ProjectHelper.load_project(name)
            return {
                "success": True,
                "message": f"Project '{name}' activated successfully",
                "project": {
                    "name": project.name if project else name,
                    "description": project.description if project else "",
                    "instructions": project.instructions if project else "",
                    "directory": ProjectHelper.get_project_directory(name)
                } if project else None
            }
        else:
            return {
                "success": False,
                "error": f"Failed to activate project '{name}'"
            }

    async def _handle_deactivate(self, context) -> Dict[str, Any]:
        """Handle project deactivation"""
        if not context:
            return {
                "success": False,
                "error": "Agent context is required for project deactivation"
            }

        current_project = ProjectHelper.get_active_project(context.agent0)
        
        if not current_project:
            return {
                "success": False,
                "error": "No project is currently active"
            }

        ProjectHelper.set_active_project(context.agent0, None)
        
        return {
            "success": True,
            "message": f"Project '{current_project}' deactivated successfully"
        }

    async def _handle_get_active(self, context) -> Dict[str, Any]:
        """Get currently active project"""
        if not context:
            return {
                "success": False,
                "error": "Agent context is required"
            }

        current_project = ProjectHelper.get_active_project(context.agent0)
        
        if not current_project:
            return {
                "success": True,
                "active_project": None,
                "message": "No project is currently active"
            }

        project = ProjectHelper.load_project(current_project)
        
        return {
            "success": True,
            "active_project": {
                "name": project.name if project else current_project,
                "description": project.description if project else "",
                "instructions": project.instructions if project else "",
                "directory": ProjectHelper.get_project_directory(current_project)
            } if project else {
                "name": current_project,
                "description": "",
                "instructions": "",
                "directory": ProjectHelper.get_project_directory(current_project)
            }
        }

    async def _handle_update(self, name: str, description: str, instructions: str) -> Dict[str, Any]:
        """Handle project update"""
        if not name:
            return {
                "success": False,
                "error": "Project name is required"
            }

        # Only update fields that were provided (not empty)
        update_description = description if description else None
        update_instructions = instructions if instructions else None

        result = ProjectHelper.update_project(
            project_name=name,
            description=update_description,
            instructions=update_instructions
        )

        return result

File: webui/js/project.js
/**
 * Project Modal JavaScript for Agent Zero
 * 
 * This module handles the project management modal interface,
 * including creating, listing, editing, and activating projects.
 * Follows Alpine.js patterns used throughout the Agent Zero UI.
 */

const projectModalProxy = {
    isOpen: false,
    isLoading: false,
    error: '',
    success: '',
    
    // Modal states
    currentView: 'list', // 'list', 'create', 'edit'
    
    // Project list data
    projects: [],
    activeProject: null,
    
    // Form data
    formData: {
        name: '',
        description: '',
        instructions: ''
    },
    
    // Edit mode data
    editingProject: null,

    /**
     * Open the project modal
     */
    async openModal() {
        this.isOpen = true;
        this.error = '';
        this.success = '';
        this.currentView = 'list';
        
        // Force update the Alpine component
        const modal = document.getElementById('projectModal');
        if (modal && window.Alpine) {
            const alpineData = window.Alpine.$data(modal);
            if (alpineData) {
                alpineData.isOpen = true;
                alpineData.error = this.error;
                alpineData.success = this.success;
                alpineData.currentView = this.currentView;
                alpineData.activeProject = this.activeProject;
                alpineData.projects = this.projects;
            } else {
                // Fallback: directly show the modal element
                modal.style.display = 'block';
            }
        }
        
        await this.loadProjects();
        await this.loadActiveProject();
    },

    /**
     * Close the modal and reset state
     */
    handleClose() {
        this.isOpen = false;
        this.error = '';
        this.success = '';
        this.currentView = 'list';
        this.resetForm();
        this.editingProject = null;
    },

    /**
     * Load all available projects from the API
     */
    async loadProjects() {
        this.isLoading = true;
        this.error = '';
        
        try {
            const response = await window.fetchApi('/project', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    action: 'list'
                })
            });
            
            const data = await response.json();
            
            if (data.success) {
                this.projects = data.projects || [];
                
                // Also update the Alpine component if it exists
                if (this._alpineComponent) {
                    this._alpineComponent.projects = this.projects;
                    this._alpineComponent.error = this.error;
                    this._alpineComponent.success = this.success;
                    this._alpineComponent.isLoading = this.isLoading;
                    this._alpineComponent.activeProject = this.activeProject;
                }
                
                // Update the header project indicator
                this.updateHeaderIndicator();
            } else {
                this.error = data.error || 'Failed to load projects';
                this.projects = [];
            }
        } catch (error) {
            this.error = 'Network error while loading projects';
            console.error('Project loading error:', error);
            if (window.toastFetchError) {
                window.toastFetchError('Error loading projects', error);
            }
            this.projects = [];
            
            // Update Alpine component for error case too
            if (this._alpineComponent) {
                this._alpineComponent.error = this.error;
                this._alpineComponent.projects = this.projects;
            }
        } finally {
            this.isLoading = false;
            
            // Update Alpine component for loading state
            if (this._alpineComponent) {
                this._alpineComponent.isLoading = this.isLoading;
            }
        }
    },

    /**
     * Load the currently active project
     */
    async loadActiveProject() {
        try {
            const response = await window.fetchApi('/project', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    action: 'get_active',
                    context: this.getCurrentContext()
                })
            });
            
            const data = await response.json();
            
            if (data.success && data.active_project) {
                this.activeProject = data.active_project;
            } else {
                this.activeProject = null;
            }
            
            // Also update the Alpine component
            if (this._alpineComponent) {
                this._alpineComponent.activeProject = this.activeProject;
            }
            
            // Update the header project indicator
            this.updateHeaderIndicator();
        } catch (error) {
            console.error('Error loading active project:', error);
            if (window.toastFetchError) {
                window.toastFetchError('Error loading active project', error);
            }
            this.activeProject = null;
        }
    },

    /**
     * Show the create project form
     */
    showCreateForm() {
        this.currentView = 'create';
        this.resetForm();
        this.error = '';
        this.success = '';
    },

    /**
     * Show the edit project form
     */
    showEditForm(project) {
        this.currentView = 'edit';
        this.editingProject = project;
        this.formData = {
            name: project.name,
            description: project.description,
            instructions: project.instructions
        };
        this.error = '';
        this.success = '';
    },

    /**
     * Go back to the project list view
     */
    showListView() {
        this.currentView = 'list';
        this.resetForm();
        this.editingProject = null;
        this.error = '';
        this.success = '';
    },

    /**
     * Reset the form data
     */
    resetForm() {
        this.formData = {
            name: '',
            description: '',
            instructions: ''
        };
    },

    /**
     * Create a new project
     */
    async createProject() {
        if (!this.formData.name.trim()) {
            this.error = 'Project name is required';
            return;
        }

        this.isLoading = true;
        this.error = '';
        this.success = '';

        try {
            const response = await window.fetchApi('/project', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    action: 'create',
                    name: this.formData.name.trim(),
                    description: this.formData.description.trim(),
                    instructions: this.formData.instructions.trim()
                })
            });

            const data = await response.json();

            if (data.success) {
                this.success = `Project '${data.project.name}' created successfully!`;
                this.resetForm();
                await this.loadProjects();
                
                // Optionally switch back to list view after a delay
                setTimeout(() => {
                    this.showListView();
                }, 2000);
            } else {
                this.error = data.error || 'Failed to create project';
            }
        } catch (error) {
            this.error = 'Network error while creating project';
            console.error('Project creation error:', error);
            if (window.toastFetchError) {
                window.toastFetchError('Error creating project', error);
            }
        } finally {
            this.isLoading = false;
        }
    },

    /**
     * Update an existing project
     */
    async updateProject() {
        if (!this.editingProject) {
            this.error = 'No project selected for editing';
            return;
        }

        this.isLoading = true;
        this.error = '';
        this.success = '';

        try {
            const response = await window.fetchApi('/project', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    action: 'update',
                    name: this.editingProject.name,
                    description: this.formData.description.trim(),
                    instructions: this.formData.instructions.trim()
                })
            });

            const data = await response.json();

            if (data.success) {
                this.success = `Project '${this.editingProject.name}' updated successfully!`;
                await this.loadProjects();
                
                // Update active project if it was the one being edited
                if (this.activeProject && this.activeProject.name === this.editingProject.name) {
                    await this.loadActiveProject();
                }
                
                // Switch back to list view after a delay
                setTimeout(() => {
                    this.showListView();
                }, 2000);
            } else {
                this.error = data.error || 'Failed to update project';
            }
        } catch (error) {
            this.error = 'Network error while updating project';
            console.error('Project update error:', error);
            if (window.toastFetchError) {
                window.toastFetchError('Error updating project', error);
            }
        } finally {
            this.isLoading = false;
        }
    },

    /**
     * Activate a project
     */
    async activateProject(projectName) {
        this.isLoading = true;
        this.error = '';
        this.success = '';

        try {
            const response = await window.fetchApi('/project', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    action: 'activate',
                    name: projectName,
                    context: this.getCurrentContext()
                })
            });

            const data = await response.json();

            if (data.success) {
                this.success = data.message || `Project '${projectName}' activated!`;
                await this.loadActiveProject();
                
                // Refresh the page to update the UI with the new project context
                setTimeout(() => {
                    window.location.reload();
                }, 1500);
            } else {
                this.error = data.error || 'Failed to activate project';
            }
        } catch (error) {
            this.error = 'Network error while activating project';
            console.error('Project activation error:', error);
            if (window.toastFetchError) {
                window.toastFetchError('Error activating project', error);
            }
        } finally {
            this.isLoading = false;
        }
    },

    /**
     * Deactivate the current project
     */
    async deactivateProject() {
        this.isLoading = true;
        this.error = '';
        this.success = '';

        try {
            const response = await window.fetchApi('/project', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    action: 'deactivate',
                    context: this.getCurrentContext()
                })
            });

            const data = await response.json();

            if (data.success) {
                this.success = data.message || 'Project deactivated!';
                await this.loadActiveProject();
                
                // Refresh the page to update the UI
                setTimeout(() => {
                    window.location.reload();
                }, 1500);
            } else {
                this.error = data.error || 'Failed to deactivate project';
            }
        } catch (error) {
            this.error = 'Network error while deactivating project';
            console.error('Project deactivation error:', error);
            if (window.toastFetchError) {
                window.toastFetchError('Error deactivating project', error);
            }
        } finally {
            this.isLoading = false;
        }
    },

    /**
     * Get the current agent context ID
     */
    getCurrentContext() {
        // Try to get context from index.js module if it's available
        if (window.getContext && typeof window.getContext === 'function') {
            return window.getContext();
        }
        
        // Try to get from Alpine store
        if (window.Alpine && window.Alpine.store('root')) {
            const store = window.Alpine.store('root');
            if (store.context) {
                return store.context;
            }
        }
        
        // Try to get from chats section (current selected chat)
        const chatsSection = document.getElementById('chats-section');
        if (chatsSection && window.Alpine) {
            const chatsAD = window.Alpine.$data(chatsSection);
            if (chatsAD && chatsAD.selected) {
                return chatsAD.selected;
            }
        }
        
        // Fallback to empty string (server will use current session context)
        return '';
    },

    /**
     * Format instructions for display (truncate if too long)
     */
    formatInstructions(instructions, maxLength = 150) {
        if (!instructions) return 'No instructions provided';
        if (instructions.length <= maxLength) return instructions;
        return instructions.substring(0, maxLength) + '...';
    },

    /**
     * Get display status for a project
     */
    getProjectStatus(project) {
        if (this.activeProject && this.activeProject.name === project.name) {
            return 'ACTIVE';
        }
        return '';
    },
    
    /**
     * Update the header project indicator
     */
    updateHeaderIndicator() {
        const indicator = document.getElementById('active-project-indicator');
        if (indicator && window.Alpine) {
            const alpineData = window.Alpine.$data(indicator);
            if (alpineData) {
                alpineData.activeProject = this.activeProject;
            }
        }
    }
};

// Wait for Alpine to be ready
document.addEventListener('alpine:init', () => {
    Alpine.data('projectModalProxy', () => ({
        init() {
            Object.assign(this, projectModalProxy);
            
            // Create a reference so we can update this component from the global object
            projectModalProxy._alpineComponent = this;
        }
    }));
});

// Keep the global assignment for backward compatibility
window.projectModalProxy = projectModalProxy;

// Export a simple function to open the modal
window.openProjectModal = function() {
    return projectModalProxy.openModal();
};

// Initialize the header indicator after a short delay
setTimeout(() => {
    if (window.projectModalProxy) {
        window.projectModalProxy.loadActiveProject();
    }
}, 1000);